test/mpi/threads/pt2pt/commperthread_perf.c

```diff
@@ -0,0 +1,240 @@
+        }
+        MPI_Waitall(WINDOW_SIZE, requests, statuses);
+    }
+
+    /* Sync */
+    MPI_Send(&sync_buf, 1, MPI_INT, 0, tid, my_comm);
+
+    if (tid == 0) {
+        MPI_Barrier(MPI_COMM_WORLD);
+        MTest_thread_barrier(runtime_num_threads);
+    } else {
+        MTest_thread_barrier(runtime_num_threads);
+    }
+
+    /* Benchmark */
+    for (win_post_i = 0; win_post_i < win_posts; win_post_i++) {
+        for (win_i = 0; win_i < WINDOW_SIZE; win_i++) {
+            MPI_Irecv(recv_buf, MESSAGE_SIZE, MPI_CHAR, 0, tid, my_comm, &requests[win_i]);
+        }
+        MPI_Waitall(WINDOW_SIZE, requests, statuses);
+    }
+
+    /* Sync */
+    MPI_Send(&sync_buf, 1, MPI_INT, 0, tid, my_comm);
+
+    free(recv_buf);
+}
+
+
+int main(int argc, char *argv[])
+{
+    int rank, size;
+    int provided;
+    double onethread_msg_rate, multithread_msg_rate, exp_multithread_msg_rate;
+    int error;
+    MPI_Info info;
+
+    MTest_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    MPI_Comm_size(MPI_COMM_WORLD, &size);
+
+    if (provided != MPI_THREAD_MULTIPLE) {
+        if (rank == 0) {
+            printf
+                ("MPI_Init_thread must return MPI_THREAD_MULTIPLE in order for this test to run.\n");
+            fflush(stdout);
+        }
+        MPI_Abort(MPI_COMM_WORLD, 1);
+    }
+
+    if (size > 2) {
+        printf("please run with exactly two processes.\n");
+        MPI_Abort(MPI_COMM_WORLD, 1);
+    }
+
+    /* Create a communicator per thread */
+    MPI_Info_create(&info);
+    MPI_Info_set(info, "mpi_assert_new_vci", "true");
+    for (int i = 0; i < NUM_THREADS; i++) {
+        MPI_Comm_dup_with_info(MPI_COMM_WORLD, info, &thread_comms[i]);
+    }
+
+    error = MTest_thread_barrier_init();
+    if (error) {
+        fprintf(stderr, "Could not create thread barrier\n");
+        MPI_Abort(MPI_COMM_WORLD, 1);
+    }
+
+    /* Run test with 1 thread */
+    runtime_num_threads = 1;
+    if (rank == 0) {
+        send_thread((void *) 0);
+    } else {
+        recv_thread((void *) 0);
+    }
+
+    onethread_msg_rate = ((double) NUM_MESSAGES / t_elapsed[0]) / 1e6;
+
+    /* Run test with multiple threads */
+    runtime_num_threads = NUM_THREADS;
+    if (rank == 0) {
+        for (int i = 1; i < NUM_THREADS; i++) {
+            MTest_Start_thread(send_thread, (void *) (long) i);
+        }
+        send_thread((void *) 0);
+    } else {
+        for (int i = 1; i < NUM_THREADS; i++) {
+            MTest_Start_thread(recv_thread, (void *) (long) i);
+        }
+        recv_thread((void *) 0);
+    }
+
+    MTest_Join_threads();
+
+    MTest_thread_barrier_free();
+
+    /* Calculate message rate with multiple threads */
+    multithread_msg_rate = 0;
+    for (int tid = 0; tid < NUM_THREADS; tid++) {
+        double my_msg_rate = ((double) NUM_MESSAGES / t_elapsed[tid]) / 1e6;
+        multithread_msg_rate += my_msg_rate;
```

Comment:
Instead of summing them up, how about compare each threads' message rate against single thread rate individually? Do `error++` for each thread that exceeds the threshold. Also print the error message include the message rate.

Reply:
Added per-thread error checking and printing out of message rate in the new version.